### PR TITLE
Nighthawk: Factories + tests

### PR DESCRIPTION
**Pull request by oschaaf**
Signed-off-by: Otto van der Schaaf <oschaaf@we-amp.com>

### CODE REVIEW

File: nighthawk/source/client/factories_impl.cc [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+SequencerPtr SequencerFactoryImpl::create(Envoy::TimeSource& time_source,
+                                          Envoy::Event::Dispatcher& dispatcher,
+                                          BenchmarkClient& benchmark_client) const {
+  StatisticFactoryImpl statistic_factory(options_);
+
+  RateLimiterPtr rate_limiter =
+      std::make_unique<LinearRateLimiter>(time_source, Frequency(options_.requests_per_second()));
+  SequencerTarget sequencer_target =
+      std::bind(&BenchmarkClient::tryStartOne, &benchmark_client, std::placeholders::_1);
```
Comment (htuch) [inline nighthawk/source/client/factories_impl.cc]: Nit: prefer lambdas, they are more readable and fewer corner cases.

File: nighthawk/source/client/factories_impl.cc [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+  SequencerPtr sequencer = std::make_unique<SequencerImpl>(
+      platform_util_, dispatcher, time_source, std::move(rate_limiter), sequencer_target,
+      statistic_factory.create(), statistic_factory.create(), options_.duration(),
+      options_.timeout());
+  return sequencer;
```
Comment (htuch) [inline nighthawk/source/client/factories_impl.cc]: Nit: Can inline with above line.

File: nighthawk/source/client/factories_impl.cc [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+StoreFactoryImpl::StoreFactoryImpl(const Options& options) : OptionBasedFactoryImpl(options) {}
+
+Envoy::Stats::StorePtr StoreFactoryImpl::create() const {
+  return std::make_unique<Envoy::Stats::IsolatedStoreImpl>();
+}
+
+StatisticFactoryImpl::StatisticFactoryImpl(const Options& options)
+    : OptionBasedFactoryImpl(options) {}
+
+StatisticPtr StatisticFactoryImpl::create() const { return std::make_unique<HdrStatistic>(); }
```
Comment (htuch) [inline nighthawk/source/client/factories_impl.cc]: Are these going to be useful long term for testing or something? They seem pretty trivial factories but I might not have the full picture.

File: nighthawk/test/factories_test.cc [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+  auto sequencer = factory.create(api_->timeSource(), dispatcher_, benchmark_client);
+  EXPECT_NE(nullptr, sequencer.get());
+}
+
+TEST_F(FactoriesTest, createStore) {
+  StoreFactoryImpl factory(options_);
+  EXPECT_NE(nullptr, factory.create().get());
+}
+
+TEST_F(FactoriesTest, createStatistic) {
```
Comment (htuch) [inline nighthawk/test/factories_test.cc]: Nit: For `TEST_F` naming of tests, you should capitalize the first letter, e.g. `CreateStatistic` (but not in the method name), to match Envoy convention. Same goes for other cases in this file.

File: nighthawk/source/client/factories_impl.cc (hunk 2 above)
Comment (oschaaf) [reply nighthawk/source/client/factories_impl.cc]: Done

File: nighthawk/test/factories_test.cc (hunk 4 above)
Comment (oschaaf) [reply nighthawk/test/factories_test.cc]: Ugh, yeah, sorry about that, fixed.

File: nighthawk/source/client/factories_impl.cc (hunk 3 above)
Comment (oschaaf) [reply nighthawk/source/client/factories_impl.cc]: Well, I was thinking that these may produce instances that could be subject to configuration in the future. E.g. hdr histogram allows specifying a precision level. And we may want to use an isolated store or not.

File: nighthawk/source/client/factories_impl.cc (hunk 1 above)
Comment (oschaaf) [reply nighthawk/source/client/factories_impl.cc]: I'll ping you on slack for this one;

**Pull request merged**